src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+
+kz:PersonShape
+    a sh:NodeShape ;
+    sh:targetClass kz:student ;
+    sh:property [
+        sh:path kz:name ;
+        sh:minLength 5
+    ] .
+```
+
+The above lines check that the `name` property of a `student` resource has a minimum length of 5 characters.
+
+## Validate RDF data against SHACL shapes
+
+We can use the `pySHACL` library to validate the RDF data against SHACL shapes. The shape constraint
+from above is read into Python code and used to validate the RDF graph.
+
+```py
+import pyshacl
+# `shapes_graph` is the SHACL constraint from above specified in Turtle format
+
+results = pyshacl.validate(
+    graph,
+    shacl_graph = shapes_graph,
+    data_graph_format = "json-ld",
+    shacl_graph_format = "ttl",
+    inplace = False,
+    serialize_report_graph = "ttl",
+    debug = False,
+)
+```
+
+Sure enough, the validation fails in this case because in our data, the name of the student `Adam`
+has only 4 characters. The validation report is output so that we can see the constraint violation.
+
+```turtle
+Validation Report
+Conforms: False
+Results (1):
+Constraint Violation in MinLengthConstraintComponent (http://www.w3.org/ns/shacl#MinLengthConstraintComponent):
+        Severity: sh:Violation
+        Source Shape: [ sh:minLength Literal("5", datatype=xsd:integer) ; sh:path kz:name ]
+        Focus Node: <http://kuzu.io/rdf-ex#Adam>
+        Value Node: Literal("Adam")
+        Result Path: kz:name
+        Message: String length not >= Literal("5", datatype=xsd:integer)
+```
+
+The violation can easily be fixed by setting the minimum length of the `name` property to a smaller value, such as 2.
+When this is done, the validation passes.
+
+```turtle
+Validation Report
+Conforms: True
+```
+
+## Visualize the RDF graph in Kùzu Explorer
+
+When building and validating RDF graphs, the ability to get visual feedback is quite useful. Kùzu
+Explorer is a web-based interface that allows you to visualize RDF graphs and query them using Cypher (no knowledge of SPARQL required).
+The instructions to launch Kùzu Explorer and connect to an existing database are shown in [the docs](https://docs.kuzudb.com/visualization/).
+
+```bash
+docker run -p 8000:8000 \
+    -v /Users/prrao/code/kuzu-rdflib/db:/database \
+    --rm kuzudb/explorer:latest
+```
+
+In a nutshell, Kùzu's RDFGraphs extension creates four distinct tables when the RDF data is loaded into the database:
+
+- `UniKG_l`: A node table that contains literals
+- `UniKG_r`: A node table that contains resources, where the primary key is the unique IRI
+- `UniKG_lt`: A relationship table that contains resource-to-literal triples
+- `UniKG_rt`: A relationship table that contains resource-to-resource triples
+
+The following Cypher query can be run to visualize the entire graph:
+
+```cypher
+MATCH (s)-[p]->(o)
+RETURN *;
+```
+
+![](/img/rdf-shacl-kuzu/demo-rdf-viz.png)
+
+As can be seen, the graph structure is identical to that shown earlier, in the conceptual representation.
```

Comment:
Change the word "conceptual" consistently here if you'll change the previous usage to "pictorial".